utils/cryptography/attenutated_credential.go

```diff
@@ -0,0 +1,86 @@
+package cryptography
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"strings"
+
+	"golang.org/x/crypto/hkdf"
+)
+
+// DecodeKeyID into the root keyID and any caveats, which will be null otherwise
+func DecodeKeyID(keyID string) (rootKeyID string, caveats map[string]string, err error) {
+	s := strings.Split(keyID, ":")
+	if len(s) != 1 && len(s) != 2 {
+		err = errors.New("invalid keyID")
+		return
+	}
+
+	rootKeyID = s[0]
+	if len(s) == 2 {
+		var caveatBytes []byte
+		var verifyCaveatBytes []byte
+		caveatBytes, err = base64.URLEncoding.DecodeString(s[1])
+		if err != nil {
+			return
+		}
+		err = json.Unmarshal(caveatBytes, &caveats)
+		if err != nil {
+			return
+		}
+
+		// Verify canonical encoding by re-marshalling
+		verifyCaveatBytes, err = json.Marshal(caveats)
+		if err != nil {
+			return
+		}
+		if !bytes.Equal(caveatBytes, verifyCaveatBytes) {
+			err = errors.New("caveats do not match canonical encoding")
+			return
+		}
+
+		return
+	}
+	caveats = nil
+	return
+}
+
+// Attenuate a root keyID and secretKey usign the provided caveats
+func Attenuate(rootKeyID string, secretKey string, caveats map[string]string) (aKeyID string, aSecretKey string, err error) {
+	_, c, err := DecodeKeyID(rootKeyID)
+	if err != nil {
+		return
+	}
+	if c != nil {
+		err = errors.New("key has already been attenutated")
+		return
+	}
+
+	if !strings.Contains(secretKey, "secret-token") {
+		err = errors.New("invalid secretKey")
+		return
+	}
+
+	caveatBytes, err := json.Marshal(caveats)
+	if err != nil {
+		return
+	}
+	km := hkdf.New(sha256.New, []byte(secretKey), nil, caveatBytes)
```

Comment:
Although the downstream dependencies are doing checks that caveats are non-empty, should we validate here or rely on downstream to validate? An HMAC over the secretkey (assuming empty caveats) doesn't do much other than obfuscate. At the very least if empty we could add in a dummy input string.

Reply:
oops! yeah we should totally do an empty check here in addition to the other ones we're doing downstream, missed it here.